cooltools/lib/common.py

```diff
@@ -221,3 +221,93 @@ def mask_cooler_bad_bins(track, bintable):
+):
+    """
+    Merge a track dataframe to a cooler bintable.
+
+    Checks that bin sizes match between a track and a cooler,
+    merges the cooler bintable with the track,
+    and optionally propagates masked regions from a cooler bintable to a track.
+
+    Parameters
+    ----------
+    track : pd.DataFrame
+        bedGraph-like track DataFrame to check
+    clr : cooler
+        cooler object to check against
+    view_df : bioframe.viewframe or None
+        Optional viewframe of regions to check for their number of bins with assigned track values.
+        If None, constructs a view_df from cooler chromsizes.
+    clr_weight_name : str
+        Name of the column in the bin table with weight
+    mask_bad_bins : bool
+        Whether to propagate null bins from cooler bintable column clr_weight_name
+        to the 'value' column of the output clr_track. Default True.
+
+    Returns
+    -------
+    clr_track
+        track dataframe that has been merged with the cooler bintable
+        and has columns ['chrom','start','end','value']
+
+
+    """
+    from .checks import is_track, is_cooler_balanced
+
+    try:
+        is_track(track, view_df=view_df, raise_errors=True)
+    except Exception as e:
+        raise ValueError("invalid input track") from e
+
+    c, s, e, v = track.columns[:4]
+    track.rename(columns={c: "chrom", s: "start", e: "end", v: "value"}, inplace=True)
+
+    track_bin_width = (track["end"] - track["start"]).median().astype(int)
+    if not (track_bin_width == clr.binsize):
+        raise ValueError(
+            "mismatch between track and cooler bin size, check track resolution"
+        )
+
+    clr_track = (
+        (clr.bins()[:])
+        .copy()
+        .merge(track, how="left", on=["chrom", "start"], suffixes=("", "_"))
+    )
+
+    if clr_weight_name:
+        try:
+            is_cooler_balanced(clr, clr_weight_name=clr_weight_name, raise_errors=True)
+        except Exception as e:
+            raise ValueError(
+                f"no column {clr_weight_name} detected in input cooler bintable"
+            ) from e
+    else:
+        clr_track[clr_weight_name] = 1.0
+
+    valid_bins = clr_track[clr_weight_name].isna() == False
+    num_valid_bins = valid_bins.sum()
+    num_assigned_bins = (clr_track["value"][valid_bins].isna() == False).sum()
+    if num_assigned_bins < 0.5 * np.sum(valid_bins):
+        warnings.warn("less than 50% of valid bins have been assigned a value")
+    if num_assigned_bins == 0:
+        raise ValueError("no track values assigned to cooler bintable")
+
+    view_df = make_cooler_view(clr) if view_df is None else view_df
+    for chrom, start, end, name in view_df[["chrom", "start", "end", "name"]].values:
+        num_assigned_bins = (
+            bioframe.select(clr_track[valid_bins], (chrom, start, end))["value"].isna()
+            == False
+        ).sum()
+        if num_assigned_bins == 0:
```

Comment:
Another question, - I already forgot if we have discussed the issue of small chromosomes e.g. `chrM` - if user-provided track lacks `chrM` it would fail here

In the previous implementation we were deriving `view_df` from track, such that `chrM`-issue would never occur - should we continue doing that ?

Reply:
just checked saddle itself - we are still doing `track_view_df`-thing ...
`merge_track_with_cooler` isn't used there just yet
this is something to consider:
1. should we add a `make_view_from_track` function to our `common` ?
2. maybe "bake in" `make_view_from_track`-functionality into `merge_track_with_cooler` ?

Reply:
I think the current approach to track validation in saddle is inconsistent with the desiderata in https://github.com/open2c/cooltools/issues/180#issuecomment-989159611

so, yeah, I was thinking we'd replace that with `merge_track_with_cooler`-- but I wanted to get a review first so as to not have the commit so big to be un-reviewable

Reply:
my thought was that if the cooler has `chrM` but that wouldn't get any assigned values, the user should specify they still want to `merge_track_with_cooler` by specifying a `view_df` lacking `chrM`

does that make sense @sergpolly ?

Reply:
another idea is to only raise that error if there were any non-na bins in the cooler to begin with in each region-- would that be better?

Reply:
understood - agreed !

Reply:
just to confirm-- agree w/ specifics in https://github.com/open2c/cooltools/issues/319 ? & non-na bins modification for the raise error?

Reply:
> my thought was that if the cooler has `chrM` but that wouldn't get any assigned values, the user should specify they still want to `merge_track_with_cooler` by specifying a `view_df` lacking `chrM`
> 
> does that make sense @sergpolly ?

I like approaches like that - they're clean and "scalable" - but I guess they're just a bit harsh on the end user, a bit unforgiving ...
on the other hand - error message looks detailed enough !
I'm all for it - not sure what other would think

Reply:
commented on #319

> another idea is to only raise that error if there were any non-na bins in the cooler bintable to begin with, for each region-- would that be better?
i.e.

:+1: this might simplify things in some scenarios